hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/client/impl/DfsClientConf.java

```diff
@@ -994,4 +983,93 @@ public String confAsString() {
+    }
+
+    /**
+     * It helps for testing to be able to disable the random factor. It should remain
+     * enabled for non-test use
+     */
+    @VisibleForTesting
+    FetchBlockLocationsRetryer(Configuration conf, boolean enableRandom) {
+      maxBlockAcquireFailures = conf.getInt(
+          DFS_CLIENT_MAX_BLOCK_ACQUIRE_FAILURES_KEY,
+          DFS_CLIENT_MAX_BLOCK_ACQUIRE_FAILURES_DEFAULT);
+      timeWindowBase = conf.getInt(
+          Retry.WINDOW_BASE_KEY,
+          Retry.WINDOW_BASE_DEFAULT);
+      timeWindowMultiplier = conf.getInt(
+          Retry.WINDOW_MULTIPLIER_KEY,
+          Retry.WINDOW_MULTIPLIER_DEFAULT);
+      timeWindowMax = conf.getInt(
+          Retry.WINDOW_MAXIMUM_KEY,
+          Retry.WINDOW_MAXIMUM_DEFAULT
+      );
+      this.enableRandom = enableRandom;
+    }
+
+    /**
+     * For tests, exposes the maximum allowed failures
+     */
+    @VisibleForTesting
+    public int getMaxBlockAcquireFailures() {
+      return maxBlockAcquireFailures;
+    }
+
+    /**
+     * Returns whether the passed number of failures is greater or equal to the maximum
+     * allowed failures.
+     */
+    public boolean isMaxFailuresExceeded(int numFailures) {
+      return numFailures >= maxBlockAcquireFailures;
+    }
+
+    /**
+     * The wait time is calculated using a grace period, a time window, and a random factor
+     * applied to that time window. With each subsequent failure, the grace period expands
+     * to the maximum value of the previous time window, and the time window upper limit expands
+     * by a constant exponential multiplier. The first retry has a grace period of 0ms.
+     *
+     * With default settings, the first failure will result in a wait time of a random number
+     * between 0 and 3000ms. The second failure will have a grace period of 3000ms, and an
+     * additional wait time of a random number between 0 and 6000ms. Subsequent failures will
+     * expand to 6000ms grace period and 0 - 9000ms, then 9000ms grace and 0 - 12000ms, etc.
+     *
+     * This behavior can be made more and less aggressive by configuring the base value (default 3000ms)
+     * and constant exponential multiplier (default 1). For example, a base of 10 and multiplier 5 could
+     * result in one very fast retry that quickly backs off in case of multiple failures. This may be useful
+     * for low latency applications. One downside with high multipliers is how quickly the backoff can get
+     * to very high numbers. One can further customize this by setting a maximum window size to cap
+     */
+    public double getWaitTime(int numFailures) {
+      double gracePeriod = backoff(numFailures);
+      double waitTimeWithRandomFactor = backoff(numFailures + 1) * getRandomFactor();
+
+      return gracePeriod + waitTimeWithRandomFactor;
+    }
+
+    private double backoff(int failures) {
+      return Math.min(timeWindowBase * Math.pow(timeWindowMultiplier, failures) * failures, timeWindowMax);
```

Comment:
If i were adding this feature as brand new, I probably wouldn't include the `* failures` here -- the base and exponential are good enough IMO. But I needed this to maintain 100% parity with the existing backoff strategy